GPS Driver (WIP)

a-gratton opened the pull request:
My initial UBLOX GPS driver, a heavily edited port of the Sparkfun Arduino library integrated with mbed. Not tested on hardware yet as the module is currently being shipped.

a-gratton commented:
Note that this is branched of of an earlier master until the makefile bugs are fixed

jetkov commented:
Hey Aidan, can you quickly summarize the changes you made to the Arduino library? Also can you add a link to the source in a header note? If the code had a license note that requires inheritance please include that as well.

azumnanji commented on lib/sensors/GPS.h:
```diff
@@ -0,0 +1,323 @@
+		int32_t altitude;		 //Number of mm above ellipsoid
+		int32_t groundSpeed;	 //mm/s
+		int32_t headingOfMotion; //Degrees * 10^-5
+	} m_nav_data{};
```
are you planning to return latitude and longitude as degrees minutes seconds or decimal? if its decimal then it should be a double

azumnanji replied:
not sure about defining structs as volatile but if you instantiate the struct with a volatile identifier and that will make each member volatile

azumnanji commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,483 @@
+	m_packetCfg.payload[5] = 0xFF;
+
+	sendSerialCommand(&m_packetCfg);
+}
```
This allows to save config of the calibration settings?

a-gratton replied:
This saves the config in the receiver's non-volatile memory

a-gratton commented:
> Hey Aidan, can you quickly summarize the changes you made to the Arduino library? Also can you add a link to the source in a header note? If the code had a license note that requires inheritance please include that as well.

I added the license and link to the header file. Regarding changes:
* I rewrote the waitForAck mechanism entirely (made it a lot simpler as not all the functionality of the original was needed) 
* Cleaned up the processing chain by removing steps specifically for NMEA messages 
* Rewrote the send command structure to only consider sending serial commands
* Added the configuration function
* Changed getter functions to just grab the most recently updated data (which is refreshed at a configurable rate) rather than polling the receiver (which has a significant delay)
* Removed a lot of extra code/constants for other message types and I2C communication (as we aren't using these)

wmmc88 commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,483 @@
+//
```
delete this

a-gratton replied:
oops IDE autogeneration

qinyang-bao commented on lib/sensors/GPS.h:
```diff
@@ -0,0 +1,339 @@
+//UBlox SAM-M8Q-0-10 GPS Sensor'
+
```
can you also add a link to the datasheet?

a-gratton replied:
Added in newest commit

qinyang-bao commented on lib/sensors/GPS.h:
```diff
@@ -0,0 +1,339 @@
+	bool m_ignoreThisPayload;
+
+	//navigation data struct
+	//TODO: check if each member needs to be volatile, or just the struct itself
```
not sure why this struct itself needs to be volatile, all its fields are changed by code explicitly...

a-gratton replied:
I'm not sure if the volatile is necessary, but the fields are being populated in an ISR context (the processing chain is attached to the uart interface serial interrupt)

qinyang-bao replied:
my bad... didn't notice it's in an ISR

qinyang-bao commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,483 @@
+}
+
+bool GPS::waitForAck() {
+	//TODO: need error checking for failed checksums, etc
```
Would be better if you could add a timeout, just so it doesn't get stuck

a-gratton replied:
Good catch, added in newest commit

qinyang-bao commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,483 @@
+
+void GPS::onByteReceive() {
+	process(m_uart.getc(), (m_activePacketBuffer == UBLOX_PACKET_PACKETCFG ? &m_packetCfg :
+							m_activePacketBuffer == UBLOX_PACKET_PACKETACK ? &m_packetAck : &m_packetBuf),
```
nested ternary operator is not very readable, but it might just be a personal preference

cindyli-13 commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,476 @@
+//I have my doubts that this works at all honestly, but I can't say for sure until I test it
+void GPS::process(uint8_t incoming, ubxPacket *incomingUBX, uint8_t requestedClass, uint8_t requestedID) {
+	if (currentSentence == NONE) {
+		if (incoming == 0xB5) //UBX binary frames start with 0xB5, aka μ
```
Can replace `0xB5` with the constant `UBX_SYNCH_1`

cindyli-13 commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,476 @@
+	}
+
+	//non-payload bytes if/else chain
+	if (((m_ubxFrameCounter == 0) && (incoming != 0xB5)) ||     //'μ'
+			((m_ubxFrameCounter == 1) && (incoming != 0x62)))  //'b'
```
Can replace `0xB5` with `UBX_SYNCH_1` and `0x62` with `UBX_SYNCH_2`

wmmc88 commented on README.md:
```diff
@@ -29,9 +29,9 @@ This repository contains:
 1. Download the development toolchain (gcc and make) and serial interface software
    
-   For Ubuntu
+   For Ubuntu (18.04 preferred)
```
can you rebase off latest master, so the diff is correct

a-gratton replied:
I'm getting a bunch of merge conflicts trying to rebase

wmmc88 commented:
Not sure what @jetkov thinks, but I think there are too many very long functions. Try to break down each function into smaller pieces that do logical things. Personally I think 15 lines of actual code is a good limit per function. It will also make things a million times easier to unit test (@jetkov is implementing this in soon). 

Another thing of note is there is a lot of branching going on. This is going to make unit testing harder and makes code harder to read. See if you can simplify it @a-gratton .

cindyli-13 commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,476 @@
+				//not ACK or requested packet, so ignore it
+				m_ignoreThisPayload = true;
+			}
+		}
```
The sparkfun library implementation stores the lengths of the LSB and MSB when `ubxFrameCounter` is at 4 and 5, was that deliberately taken out here?

cindyli-13 commented on lib/sensors/GPS.cpp:
```diff
@@ -0,0 +1,476 @@
+	if (incomingUBX->counter < incomingUBX->len + 4) {
+		addToChecksum(incoming);
+	}
+	if (incomingUBX->counter == 0) {
+		incomingUBX->cls = incoming;
+	} else if (incomingUBX->counter == 1) {
+		incomingUBX->id = incoming;
+	} else if (incomingUBX->counter == 2) {//Len LSB
+		incomingUBX->len = incoming;
+	} else if (incomingUBX->counter == 3) { //Len MSB
+		incomingUBX->len |= incoming << 8;
+	} else if (incomingUBX->counter == incomingUBX->len + 4) { //ChecksumA
+		incomingUBX->checksumA = incoming;
+	} else if (incomingUBX->counter == incomingUBX->len + 5) { //ChecksumB, last byte of sentence trasmission
+		incomingUBX->checksumB = incoming;
```
Can replace with a switch statement
